Extension Types

nikkomiu opened the pull request:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [ ] new service onboarding
      - [ ] new API version
      - [ ] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month.
  4. If updating an existing version, please select the specific langauge SDKs and CLIs that must be refreshed after the swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [ ] No refresh required for updates in this PR

### Contribution checklist:
- [ ] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [ ] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [ ] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist 

> **Applicability**: :warning: 
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs 
> - Adding new properties 
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [ ] Check this box if any of the following apply to the PR so that label "WaitForARMFeedback" will be added automatically to begin ARM API Review. Failure to comply may result in delays to the manifest.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[ ] To review changes efficiently, ensure you copy the existing version into the new directory structure for first commit and then push new changes, including version updates, in separate commits.

- [ ] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If any of the following scenarios apply to the PR, request approval from the Breaking Change Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 
- [ ] Removing API(s) in a stable version
- [ ] Removing properties in a stable version
- [ ] Removing API version(s) in a stable version
- [ ] Updating API in a stable or public preview version with Breaking Change Validation errors
- [ ] Updating API(s) in public preview over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

NarayanThiru commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/stable/2023-05-01/extensions.json:
```diff
@@ -402,6 +402,160 @@
+          {
+            "name": "releaseTrain",
+            "in": "query",
+            "description": "Release train",
```
Is ReleaseTrain a parameter to this api?  Do you plan on filtering based on releaseTrain?  If so, explain this in the description of this parameter.

NarayanThiru commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/stable/2023-05-01/extensions.json:
```diff
@@ -402,6 +402,160 @@
+          },
+          {
+            "name": "publisherId",
+            "in": "query",
```
Similar to above question.  Pl. explain usage in the description.

NarayanThiru commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/stable/2023-05-01/extensions.json:
```diff
@@ -402,6 +402,160 @@
+            "type": "string"
+          },
+          {
+            "name": "offerId",
```
Will OfferId be used as a parameter?  Explain in description.  Make sure to get this reviewed by the MarketPlace crew

NarayanThiru commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/stable/2023-05-01/extensions.json:
```diff
@@ -402,6 +402,160 @@
+          },
+          {
+            "name": "planId",
+            "in": "query",
```
Similar to above question.  Pl. explain usage in the description, if this will be used

NarayanThiru commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/stable/2023-05-01/extensions.json:
```diff
@@ -402,6 +402,160 @@
+    "/subscriptions/{subscriptionId}/providers/Microsoft.KubernetesConfiguration/locations/{location}/extensionTypes": {
+      "get": {
+        "tags": [
+          "Extensions",
```
Replace Extensions tag with ExtensionTypes, as this is not extensions resource.  Update ClusterExtensions with ClusterExtensionTypes.

NarayanThiru commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/stable/2023-05-01/extensions.json:
```diff
@@ -402,6 +402,160 @@
+    "/subscriptions/{subscriptionId}/providers/Microsoft.KubernetesConfiguration/locations/{location}/extensionTypes/{extensionName}": {
+      "get": {
+        "tags": [
+          "Extensions",
```
Replace Extensions tag with ExtensionTypes, as this is not extensions resource.  Update ClusterExtensions with ClusterExtensionTypes.

NarayanThiru commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/stable/2023-05-01/extensions.json:
```diff
@@ -402,6 +402,160 @@
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/providers/Microsoft.KubernetesConfiguration/locations/{location}/extensionTypes/{extensionName}": {
```
It should  be extensionTypeName

NarayanThiru commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/stable/2023-05-01/extensions.json:
```diff
@@ -402,6 +402,160 @@
+            "$ref": "../../../../../common-types/resource-management/v3/types.json#/parameters/LocationParameter"
+          },
+          {
+            "$ref": "#/parameters/ExtensionNameParameter"
```
This must be ExtensionTypeNameParameter

NarayanThiru commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/stable/2023-05-01/extensions.json:
```diff
@@ -402,6 +402,160 @@
+        }
+      }
+    },
+    "/subcriptions/{subscriptionId}/providers/Microsoft.KubernetesConfiguration/locations/{location}/extensionTypes/{extensionName}/versions": {
```
This must be extensionTypeName

NarayanThiru commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/stable/2023-05-01/extensions.json:
```diff
@@ -402,6 +402,160 @@
+    "/subcriptions/{subscriptionId}/providers/Microsoft.KubernetesConfiguration/locations/{location}/extensionTypes/{extensionName}/versions": {
+      "get": {
+        "tags": [
+          "Extensions",
```
Replace Extensions tag with ExtensionTypes, as this is not extensions resource.  Update ClusterExtensions with ClusterExtensionTypes.

NarayanThiru commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/stable/2023-05-01/extensions.json:
```diff
@@ -402,6 +402,160 @@
+            "$ref": "../../../../../common-types/resource-management/v3/types.json#/parameters/LocationParameter"
+          },
+          {
+            "$ref": "#/parameters/ExtensionNameParameter"
```
This must be ExtensionTypeNameParameter

NarayanThiru commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/stable/2023-05-01/extensions.json:
```diff
@@ -692,6 +846,142 @@
+              "description": "The name of the extension type.",
+              "type": "string"
+            },
+            "type": {
```
Do we have a type for extensionType?

nikkomiu replied:
I didn't see a type for ExtensionType

nikkomiu commented:
@NarayanThiru I updated based on your comments. It's going to be a bit of a pain because I also switched it to its own file since it seems to make more sense that way. I can always move it back if you don't like it this way though.

kulalking commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/stable/2023-05-01/extensionTypes.json:
```diff
@@ -0,0 +1,340 @@
+            "description": "Filter results by Plan ID of a marketplace extension",
+            "required": false,
+            "type": "string"
+          }
```
we dont need this right ?

nikkomiu replied:
I thought they said that they still wanted it to be on there. We can ask them later today to verify.

kulalking replied:
they? is that Arif? lets sync

kulalking commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/stable/2023-05-01/extensionTypes.json:
```diff
@@ -0,0 +1,340 @@
+          "200": {
+            "description": "OK",
+            "schema": {
+              "$ref": "#/definitions/ExtensionTypesList"
```
>ExtensionTypesList [](http://example.com/codeflow?start=37&length=18)

can you please add this also part of the review. this is what  even then ARM swagger reviewers will ask you to do. it helps to review faster.

kulalking commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/stable/2023-05-01/extensionTypes.json:
```diff
@@ -0,0 +1,340 @@
+          "items": {
+            "$ref": "#/definitions/ExtensionType"
+          },
+          "description": "List of Extension Types within a Kubernetes cluster."
```
>List of Extension Types within a Kubernetes cluster. [](http://example.com/codeflow?start=26&length=52)

List of ext types (please remove within a kub cluster)

kulalking commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/stable/2023-05-01/extensionTypes.json:
```diff
@@ -0,0 +1,340 @@
+              "description": "The default scope of the extension type.",
+              "type": "string"
+            },
+            "properties": {
```
>properties [](http://example.com/codeflow?start=13&length=10)

Even these details we are asking publishers to provide is it?

nikkomiu replied:
Yea, I believe so. We can double check later today.

kulalking commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/stable/2023-05-01/extensionTypes.json:
```diff
@@ -0,0 +1,340 @@
+              "items": {
+                "type": "object",
+                "properties": {
+                  "key": {
```
>key" [](http://example.com/codeflow?start=19&length=4)

do we need to have type also here ?

nikkomiu replied:
I can add this if you'd like. I guess even if it doesn't get used (initially) it probably won't be a big deal...and we could just default everything to string 😛

nikkomiu replied:
I went back and looked at my notes from the first meeting we had. They didn't want to have a type property. They thought it would be overkill.

kulalking commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/stable/2023-05-01/extensionTypes.json:
```diff
@@ -0,0 +1,340 @@
+              "items": {
+                "type": "object",
+                "properties": {
+                  "key": {
```
>"key" [](http://example.com/codeflow?start=18&length=5)

do we need type also here?

kulalking commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/stable/2023-05-01/extensionTypes.json:
```diff
@@ -0,0 +1,340 @@
+                }
+              }
+            },
+            "id": {
```
> "id": [](http://example.com/codeflow?start=11&length=6)

what is this? i think we can remove this unless you have something in your mind.

kulalking commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/stable/2023-05-01/extensionTypes.json:
```diff
@@ -0,0 +1,340 @@
+              "description": "The name of the extension type.",
+              "type": "string"
+            },
+            "type": {
```
>"type": [](http://example.com/codeflow?start=12&length=7)

what is type represents ?

nikkomiu replied:
Obviously, type represents the type of the extension type 😝. I removed it. Not sure why it was there either...

kulalking commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/stable/2023-05-01/extensionTypes.json:
```diff
@@ -0,0 +1,340 @@
+    }
+  },
+  "definitions": {
+    "ExtensionType": {
```
>ExtensionType [](http://example.com/codeflow?start=5&length=13)

we need Publisher, Description, IsSystemExtension(check with Arif is this required),  planinfo (check with arif)

kulalking replied:
>ExtensionType [](http://example.com/codeflow?start=5&length=13)

do we need to show list of supported regions as well ?

kulalking reviewed (changes requested):
:clock1:

kulalking commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/stable/2023-05-01/extensionTypes.json:
```diff
@@ -0,0 +1,340 @@
+        "operationId": "ExtensionTypes_LocationList",
+        "x-ms-examples": {
+          "List Extension Types": {
+            "$ref": "./examples/ListExtensionTypesByLocation.json"
```
>ListExtensionTypesByLocation [](http://example.com/codeflow?start=32&length=28)

can you please add this also part of the review. it helps to review even for ARM team.

kulalking commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/stable/2023-05-01/extensionTypes.json:
```diff
@@ -0,0 +1,340 @@
+            "$ref": "./examples/ListExtensionTypesByLocation.json"
+          }
+        },
+        "parameters": [
```
>parameters [](http://example.com/codeflow?start=9&length=10)

apiversion parameter is missing here.

kulalking commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/stable/2023-05-01/extensionTypes.json:
```diff
@@ -0,0 +1,340 @@
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/providers/Microsoft.KubernetesConfiguration/locations/{location}/extensionTypes/{extensionTypeName}": {
```
>locations/{location} [](http://example.com/codeflow?start=81&length=20)

why do we need this ?

nikkomiu replied:
This is going to be a "location based" route for the extension types. There was another one that was discussed that was "cluster-based" but I want to see if that one is actually going to be relevant since extension types isn't actually cluster based.

kulalking commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/stable/2023-05-01/extensionTypes.json:
```diff
@@ -0,0 +1,340 @@
+        }
+      }
+    },
+    "/subcriptions/{subscriptionId}/providers/Microsoft.KubernetesConfiguration/locations/{location}/extensionTypes/{extensionTypeName}/versions": {
```
>locations/{location} [](http://example.com/codeflow?start=80&length=20)

can we move this to query parameter, we can discuss this with PM if required

nikkomiu replied:
So you're saying that you want the route to be /subcriptions/{subscriptionId}/providers/Microsoft.KubernetesConfiguration/extensionTypes (and all routes sub-routes to that)?

kulalking commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/stable/2023-05-01/extensionTypes.json:
```diff
@@ -0,0 +1,340 @@
+        }
+      }
+    },
+    "/subcriptions/{subscriptionId}/providers/Microsoft.KubernetesConfiguration/locations/{location}/extensionTypes/{extensionTypeName}/versions": {
```
>versions [](http://example.com/codeflow?start=136&length=8)

can we please add the section for query parameters and add the all the supported query parameters, release train , location, plan info check with arif

nikkomiu replied:
Yea, they should all be there. Versions doesn't have query params but the other routes where they're applicable have them listed. https://github.com/jagpreetstamber/azure-rest-api-specs/pull/4/files/c0f3e9a24d61150f6279e04ffb591a1c6bd4cb17#diff-7ae1ee61c9a7e9c57b4668dbff49e25f84c1c6fafc5417eaae270190144386e2R57-R84

kulalking commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/stable/2023-05-01/extensionTypes.json:
```diff
@@ -0,0 +1,340 @@
+{
+  "swagger": "2.0",
```
>swagger [](http://example.com/codeflow?start=3&length=7)

you have missed the get extension types based on cluster types, is that dropped ?

nikkomiu replied:
Added those routes

kulalking commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/stable/2023-05-01/extensionTypes.json:
```diff
@@ -0,0 +1,476 @@
+    }
+  },
+  "paths": {
+    "/subscriptions/{subscriptionId}/providers/Microsoft.KubernetesConfiguration/locations/{location}/extensionTypes": {
```
>locations/{location} [](http://example.com/codeflow?start=81&length=20)

can we move it to query parameter

kulalking commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/stable/2023-05-01/extensionTypes.json:
```diff
@@ -0,0 +1,476 @@
+            "$ref": "../../../../../common-types/resource-management/v2/types.json#/parameters/SubscriptionIdParameter"
+          },
+          {
+            "$ref": "../../../../../common-types/resource-management/v3/types.json#/parameters/LocationParameter"
```
>LocationParameter [](http://example.com/codeflow?start=95&length=17)

release train

kulalking commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/stable/2023-05-01/extensionTypes.json:
```diff
@@ -0,0 +1,476 @@
+            "description": "Filter results by release train",
+            "required": false,
+            "type": "string"
+          },
```
not required

kulalking commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/stable/2023-05-01/extensionTypes.json:
```diff
@@ -0,0 +1,497 @@
+            "description": "Filter results by release train",
+            "required": false,
+            "type": "string"
+          },
```
we don't need this.

kulalking commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/stable/2023-05-01/extensionTypes.json:
```diff
@@ -0,0 +1,497 @@
+            "description": "Filter results by release train",
+            "required": false,
+            "type": "string"
+          },
```
we dont need this as well

NarayanThiru commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/stable/2023-05-01/extensionTypes.json:
```diff
@@ -0,0 +1,483 @@
+  "swagger": "2.0",
+  "info": {
+    "version": "2023-05-01",
+    "title": "ExtensionsClient",
```
>ExtensionsClient [](http://example.com/codeflow?start=14&length=16)

Change this to ExtensionTypesClient

NarayanThiru commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/stable/2023-05-01/extensionTypes.json:
```diff
@@ -0,0 +1,483 @@
+          {
+            "name": "publisherId",
+            "in": "query",
+            "description": "Filter results by Publisher ID of a marketplace extension",
```
>extension [](http://example.com/codeflow?start=76&length=9)

Extension Type

NarayanThiru commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/stable/2023-05-01/extensionTypes.json:
```diff
@@ -0,0 +1,483 @@
+        }
+      }
+    },
+    "ExtensionTypeVersion": {
```
>ExtensionTypeVersion [](http://example.com/codeflow?start=5&length=20)

This is misleading - as it contains ReleaseTrain as well.  May be a name like ExtensionTypeVersionsByReleaseTrain would be better, no?

NarayanThiru commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/stable/2023-05-01/extensionTypes.json:
```diff
@@ -0,0 +1,483 @@
+          "ExtensionTypes",
+          "ClusterExtensionTypes"
+        ],
+        "description": "List Extension Types for the cluster.",
```
>List Extension Types for the cluster [](http://example.com/codeflow?start=24&length=36)

Is this ExtensionTypes of extensions 'installed' in the cluster or that 'can be installed' in the cluster?  If latter, how do you determine what extensions can be installed?  Explain that in the description, so it is clear what the intent of this api is.  This will go in the public documentation of our spec!

NarayanThiru commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/stable/2023-05-01/extensionTypes.json:
```diff
@@ -0,0 +1,483 @@
+          {
+            "name": "offerId",
+            "in": "query",
+            "description": "Filter results by Offer ID of a marketplace extension",
```
>extension [](http://example.com/codeflow?start=72&length=9)

Extension Type

NarayanThiru commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/stable/2023-05-01/extensionTypes.json:
```diff
@@ -0,0 +1,483 @@
+      }
+    },
+    "/subcriptions/{subscriptionId}/providers/Microsoft.KubernetesConfiguration/locations/{location}/extensionTypes/{extensionTypeName}/versions": {
+      "get": {
```
We will need a GET of Version as well.  This will used to check if a version exists or not - we can't expect the client to do a list and go through it to check if it exists

NarayanThiru commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/stable/2023-05-01/extensionTypes.json:
```diff
@@ -0,0 +1,483 @@
+      }
+    }
+  },
+  "definitions": {
```
>definitions [](http://example.com/codeflow?start=3&length=11)

Move these types under common

NarayanThiru commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/stable/2023-05-01/extensionTypes.json:
```diff
@@ -0,0 +1,483 @@
+            "properties": {
+              "publisherId": {
+                "type": "string",
+                "description": "Marketplace Extension Type Publisher ID."
```
>Marketplace Extension Type Publisher ID [](http://example.com/codeflow?start=32&length=39)

Reword this to "Publisher Id of the Marketplace Extension"

NarayanThiru commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/stable/2023-05-01/extensionTypes.json:
```diff
@@ -0,0 +1,483 @@
+            "type": "string",
+            "description": "Name of the publisher for the Extension Type"
+          },
+          "planInfo": {
```
>planInfo [](http://example.com/codeflow?start=11&length=8)

Add a description "Only for Marketplace Extension Types"

NarayanThiru commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/stable/2023-05-01/extensionTypes.json:
```diff
@@ -0,0 +1,483 @@
+            }
+          },
+          "supportedClusterTypes": {
+            "description": "The list of cluster types of the extension type.",
```
>The list of cluster types of the extension type [](http://example.com/codeflow?start=28&length=47)

Reword to "Cluster Types supported for this Extension Type"

NarayanThiru commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/stable/2023-05-01/extensionTypes.json:
```diff
@@ -0,0 +1,483 @@
+          },
+          "supportedScopes": {
+            "defaultScope": {
+              "description": "The default scope of the extension type.",
```
>extension type [](http://example.com/codeflow?start=55&length=14)

Add "This scope will be used if the user does not provide a scope while creating an extension"

NarayanThiru commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/stable/2023-05-01/extensionTypes.json:
```diff
@@ -0,0 +1,483 @@
+          "type": "string"
+        },
+        "versions": {
+          "description": "The list of versions of the extension type.",
```
>extension type. [](http://example.com/codeflow?start=54&length=15)

expand this to say "...for this Release Train"

NarayanThiru commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/stable/2023-05-01/extensionTypes.json:
```diff
@@ -0,0 +1,483 @@
+          "items": {
+            "$ref": "#/definitions/ExtensionTypeVersion"
+          },
+          "description": "List of Extension Type Versions for an Extension Type."
```
>Type Versions for an Extension Type. [](http://example.com/codeflow?start=44&length=36)

You need to clarify that this list is Versions per ReleaseTrain

nikkomiu commented:
Add Query Parameter to Location-based routing for `extensionType`

kulalking commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/preview/2023-05-01-preview/extensionTypes.json:
```diff
@@ -0,0 +1,725 @@
+            "description": "Filter results by extension type name",
+            "required": false,
+            "type": "string"
+          },
```
remove

kulalking commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/preview/2023-05-01-preview/extensionTypes.json:
```diff
@@ -0,0 +1,725 @@
+            "description": "Filter results by the cluster type for extension types",
+            "required": false,
+            "type": "string"
+          },
```
remove this

kulalking commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/preview/2023-05-01-preview/extensionTypes.json:
```diff
@@ -0,0 +1,711 @@
+            "$ref": "../../common/2022-03-01/parameters.json#/parameters/ClusterNameParameter"
+          },
+          {
+            "name": "publisherId",
```
>publisherId [](http://example.com/codeflow?start=21&length=11)

plan id ? question

NarayanThiru commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/preview/2023-05-01-preview/extensionTypes.json:
```diff
@@ -0,0 +1,749 @@
+        }
+      }
+    },
+    "/subcriptions/{subscriptionId}/providers/Microsoft.KubernetesConfiguration/locations/{location}/extensionTypes/{extensionTypeName}/versions": {
```
>"/subcriptions [](http://example.com/codeflow?start=4&length=14)

We'll need a new api that gives the latest version for a given ExtensionType+ReleaseTrain+ClusterType

ponatara commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/preview/2023-05-01-preview/examples/GetExtensionTypeVersion.json:
```diff
@@ -0,0 +1,56 @@
+                "unsupportedVersions": [">1.22"]
+              }
+            ],
+            "appliances": [
```
Can you confirm that we'll filter to show only the unsupported versions for a cluster type if the request is in context of a cluster type?

nikkomiu replied:
Yea, that's what we'll do. But if they don't specify a cluster type we'll show all unsupported versions for all cluster types.

ponatara commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/preview/2023-05-01-preview/examples/GetExtensionType.json:
```diff
@@ -0,0 +1,41 @@
+      "body": {
+        "properties": {
+          "isSystemExtension": false,
+          "shouldCreateIdentity": true,
```
Did you get a chance to check with Anubhav on why we need to include ShouldCreateIdentity?

nikkomiu replied:
Yea, this is something that partner teams were looking for when using the CLI so they could determine if they're going to need the identity. Right now they have to do some weird hacky things to get it working properly and he'd really like this property to be added for the API.

ponatara replied:
I spoke to Anubhav. Can we rename this property to be more user-friendly -"isManagedIdentityRequired": true 
Summarizing discussion for everyone - Some solutions need a MI (like Azure backup, Azure Monitor). ARM creates the identity in user's tenant. Our RP cannot do that, so this intent has to come through the ARM API (which we have today). Exposing this as part of this API will make it apparent to customers that some extensions need identity and customers can pass it if using REST API

nikkomiu replied:
Done

RamyasreeChakka commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/preview/2023-05-01-preview/extensionTypes.json:
```diff
@@ -0,0 +1,784 @@
+          {
+            "name": "offerId",
+            "in": "query",
+            "description": "Filter results by Offer ID of a marketplace extension type",
```
>Filter results by Offer ID of a marketplace extension type [](http://example.com/codeflow?start=28&length=58)

Please use below text
Filter results by Offer ID or Product ID of a marketplace extension type"

nikkomiu replied:
This doesn't do both. This field only filters for the Offer ID. The productId field filters for the Product ID.

RamyasreeChakka commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/preview/2023-05-01-preview/extensionTypes.json:
```diff
@@ -0,0 +1,784 @@
+          {
+            "name": "offerId",
+            "in": "query",
+            "description": "Filter results by Offer ID of a marketplace extension type",
```
>Filter results by Offer ID of a marketplace extension type [](http://example.com/codeflow?start=28&length=58)

Filter results by Offer ID or Product ID of a marketplace extension type"

nikkomiu commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/preview/2023-05-01-preview/extensionTypes.json:
```diff
@@ -0,0 +1,784 @@
+              },
+              "offerId": {
+                "type": "string",
+                "description": "Offer ID of the Marketplace Extension Type."
```
I only need one of these. When they're the same thing I just bulk update them all.

RamyasreeChakka commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/preview/2023-05-01-preview/examples/GetExtensionType.json:
```diff
@@ -0,0 +1,41 @@
+      "headers": {},
+      "description": "Details of the extension type",
+      "body": {
+        "properties": {
```
When you submit the pull request to ARM rest api specs repo, I think you may run into a check which validates the GET reponses follow the ARM resource model. 

In this model, I see you have properties but do not have top level properties like id, name, type.

nikkomiu replied:
They're in there now

RamyasreeChakka commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/preview/2023-05-01-preview/examples/GetExtensionTypeByLocation.json:
```diff
@@ -0,0 +1,40 @@
+      "description": "Details of the extension type",
+      "body": {
+        "value": {
+          "properties": {
```
Why does this response contain 'value'? 
Get on a specific extension type should return only single object similar to GetExtensionType.json example

RamyasreeChakka commented on specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/preview/2023-05-01-preview/examples/GetExtensionTypeVersionByLocation.json:
```diff
@@ -0,0 +1,55 @@
+      "description": "Versions for the Extension Type",
+      "body": {
+        "value": {
+          "version": "1.1.0",
```
Use 'value' property for only list responses i.e. when the URI has odd number of segments.

RamyasreeChakka reviewed (approved):
:shipit:
